src/app/qgisapp.cpp

```diff
@@ -3240,16 +3240,9 @@ bool QgisApp::addVectorLayers( const QStringList &theLayerQStringList, const QSt
         //set friendly name for datasources with only one layer
         QStringList sublayers = layer->dataProvider()->subLayers();
         QStringList elements = sublayers.at( 0 ).split( ':' );
-        if ( layer->storageType() != "GeoJSON" )
-        {
-          while ( elements.size() > 4 )
-          {
-            elements[1] += ':' + elements[2];
-            elements.removeAt( 2 );
-          }
 
-          layer->setLayerName( elements.at( 1 ) );
-        }
+        layer->setLayerName( QString( "%1 %2 %3" ).arg( layer->name() ).arg( elements.at( 1 ) ).arg( elements.at( 3 ) ) );
```

Comment:
is this safe? will elements always have at least 4 items?

(also, FYI, it's better to use the multiargument .arg( QString, QString, ... ) method instead of chaining .arg()'s. It's faster and safer if any of the strings contain "%" characters)